src/pure/cmd_analysis.ml

```diff
@@ -52,6 +52,48 @@ and qidents_of_p_term (term : Syntax.p_term) =
   | Syntax.P_Wrap term -> qidents_of_p_term term
   | Syntax.P_Expl term -> qidents_of_p_term term
 
+and qidents_of_rw_patt (rwpat : Syntax.p_rw_patt) =
+  match rwpat with
+  | Syntax.P_rw_Term pt -> qidents_of_p_term pt
+  | Syntax.P_rw_InTerm pt -> qidents_of_p_term pt
+  | Syntax.P_rw_InIdInTerm (_, pt) -> qidents_of_p_term pt
+  | Syntax.P_rw_IdInTerm (_, pt) -> qidents_of_p_term pt
+  | Syntax.P_rw_TermInIdInTerm (pt1, _, pt2) ->
+      (qidents_of_p_term pt1) @ (qidents_of_p_term pt2)
+  | Syntax.P_rw_TermAsIdInTerm (pt1, _, pt2) ->
+      (qidents_of_p_term pt1) @ (qidents_of_p_term pt2)
+
+and qidents_of_p_assertion (pasrtn : Syntax.p_assertion) =
+  match pasrtn with
+  | Syntax.P_assert_typing (pt, _) -> qidents_of_p_term pt
+  | Syntax.P_assert_conv (pt1, pt2) ->
+    qidents_of_p_term pt1 @ qidents_of_p_term pt2
+
+and qidents_of_p_query (pq : Syntax.p_query) =
+  match pq.elt with
+  | Syntax.P_query_assert (_, pasrt)-> qidents_of_p_assertion pasrt
+  | Syntax.P_query_infer (pt, _) -> qidents_of_p_term pt
+  | Syntax.P_query_normalize (pt, _) -> qidents_of_p_term pt
+  | Syntax.P_query_print qidnt_opt ->
+    (match qidnt_opt with Some qidnt -> [qidnt] | None -> [])
+  | _ -> []
+
+
+and qidents_of_p_tactic (term: Syntax.p_tactic) =
+  match term.elt with
+  | Syntax.P_tac_refine pt -> qidents_of_p_term pt
+  | Syntax.P_tac_apply pt -> qidents_of_p_term pt
+  | Syntax.P_tac_rewrite (_, rwpat_lo, pt) ->
+    let rwqids =
+      match rwpat_lo with
+      | Some rwpat -> qidents_of_rw_patt rwpat.elt
+      | None -> []
+    in
```

Comment:
You could factorize these lines, lines 139-143 and line 138 by using List.map_default:
- lines 88-91 replaced by `List.map_default qidents_of_rw_patt [] rwpat.elt`
- lines 140-143 replaced by `List.map_default (List.concat_map qidents_of_p_tactic) [] ptac_list`
- line 138 removed
- some_or_empty replaced by `List.map_default (fun x -> [x]) []`